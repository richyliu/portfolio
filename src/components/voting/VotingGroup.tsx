import React, { ReactElement } from "react"
import { Form } from "react-bootstrap"

interface VotingGroupProps {
  /**
   * The title of the group.
   */
  title?: string;
  /**
   * The description of the group.
   */
  description?: string;
  /**
   * The options to render. Each option should have a label property dictating what will be
   * shown to the user, and a value property, which will be referenced in events.
   */
  options: {
    /**
     * TEST Label
     */
    label: string;
    value: string | number;
  }[];
  /**
   * An array of values that should be selected.
   */
  value?: (string | number)[];
  /**
   * A function to be called each time the value of an option changes.
   * @param value - The `value` of the option that was changed.
   * @param state - Whether or not the checkbox should now be checked (after the change).
   */
  onSelectChange: (value: string | number, state: boolean) => void;
}

/*
 * TODO
 * Add typedoc and see how that works Maybe even write some way to link typedoc with storybook? Probably not that hard.
 * Re setup storybook because we don't use storybook for gatsby pages, only components.
 */
const VotingGroup = ({
  title,
  description,
  options,
  value,
  onSelectChange,
}: VotingGroupProps): ReactElement => {
  return (
    <div>
      <h4 className="mb-1">{title}</h4>
      <p className="text-muted mb-2 mt-1" hidden={!description}>
        {description}
      </p>
      {options.map((opt, i) => {
        const checked = value?.indexOf(opt.value + "") > -1
        return (
          <Form.Check
            key={i}
            custom
            id={`check-${i}-${typeof opt.value}-${opt.value}`} // required for react bootstrap
            label={opt.label}
            checked={checked}
            onChange={(): void => onSelectChange(opt.value, !checked)}
          />
        )
      })}
    </div>
  )
}
// eslint-disable-next-line no-undef
export {VotingGroup, VotingGroupProps}
export default VotingGroup
