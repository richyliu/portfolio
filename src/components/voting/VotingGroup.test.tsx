import * as React from "react"
import { mount, shallow } from "enzyme"
import toJson from "enzyme-to-json"
import VotingGroup, { VotingGroupProps } from "./VotingGroup"

const defaultProps: Partial<VotingGroupProps> = {
  title: "Title",
  description: "A nice description can go here.",
}

describe("VotingGroup calls onSelectChanged with the proper argument", () => {
  test("with number values", () => {
    const onSelectChange = jest.fn()
    const props: VotingGroupProps = {
      ...defaultProps,
      onSelectChange,
      options: [
        {
          label: "Option One",
          value: 1,
        },
        {
          label: "Option Two",
          value: 2,
        },
        {
          label: "Option Three",
          value: 3,
        },
      ],
    }

    // subject under test
    const sut = shallow(<VotingGroup {...props} />)
    // console.log(sut);
    sut
      .find("#check-0-number-1")
      .simulate("change", { target: { value: "true" } })
    expect(onSelectChange).toHaveBeenNthCalledWith(1, 1)
    sut
      .find("#check-1-number-2")
      .simulate("change", { target: { value: "true" } })
    expect(onSelectChange).toHaveBeenNthCalledWith(2, 2)
    sut
      .find("#check-2-number-3")
      .simulate("change", { target: { value: "true" } })
    expect(onSelectChange).toHaveBeenNthCalledWith(3, 3)
  })
  test("with string values", () => {
    const onSelectChange = jest.fn()
    const props: VotingGroupProps = {
      ...defaultProps,
      onSelectChange,
      options: [
        {
          label: "Option One",
          value: "1",
        },
        {
          label: "Option Two",
          value: "two",
        },
        {
          label: "Option Three",
          value: "thiisareallylongstringvalue",
        },
      ],
    }

    const sut = shallow(<VotingGroup {...props} />)
    // console.log(sut);
    sut
      .find("#check-0-string-1")
      .simulate("change", { target: { value: "true" } })
    expect(onSelectChange).toHaveBeenNthCalledWith(1, "1")
    sut
      .find("#check-1-string-two")
      .simulate("change", { target: { value: "true" } })
    expect(onSelectChange).toHaveBeenNthCalledWith(2, "two")
    sut
      .find("#check-2-string-thiisareallylongstringvalue")
      .simulate("change", { target: { value: "true" } })
    expect(onSelectChange).toHaveBeenNthCalledWith(
      3,
      "thiisareallylongstringvalue"
    )
  })
})
describe("It matches the snapshot", () => {
  test("with title and description", () => {
    const props: VotingGroupProps = {
      ...defaultProps,
      title: undefined,
      description: undefined,
      onSelectChange: () => {
        /* noop */
      },
      options: [
        {
          label: "Option One",
          value: "1",
        },
        {
          label: "Option Two",
          value: 2,
        },
        {
          label: "Option Three",
          value: "thiisareallylongstringvalue",
        },
      ],
      value: ["1", 2],
    }

    const sut = mount(<VotingGroup {...props} />)
    expect(toJson(sut)).toMatchSnapshot()
    // expect(true).toBe(true);
    sut.unmount()
  })
  test("with no title or description", () => {
    const props: VotingGroupProps = {
      ...defaultProps,
      title: undefined,
      description: undefined,
      onSelectChange: () => {
        /* noop */
      },
      options: [
        {
          label: "Option One",
          value: "1",
        },
        {
          label: "Option Two",
          value: 2,
        },
        {
          label: "Option Three",
          value: "thiisareallylongstringvalue",
        },
      ],
      value: ["1", 2],
    }

    const sut = mount(<VotingGroup {...props} />)
    expect(toJson(sut)).toMatchSnapshot()
    // expect(true).toBe(true);
    sut.unmount()
  })
})
