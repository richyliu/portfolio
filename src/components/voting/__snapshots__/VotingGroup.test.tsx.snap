// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It matches the snapshot with no title or description 1`] = `
<Component
  onSelectChange={[Function]}
  options={
    Array [
      Object {
        "label": "Option One",
        "value": "1",
      },
      Object {
        "label": "Option Two",
        "value": 2,
      },
      Object {
        "label": "Option Three",
        "value": "thiisareallylongstringvalue",
      },
    ]
  }
  value={
    Array [
      "1",
      2,
    ]
  }
>
  <div>
    <h4
      className="mb-1"
    />
    <p
      className="text-muted mb-2 mt-1"
      hidden={true}
    />
    <FormCheck
      checked={true}
      custom={true}
      disabled={false}
      id="check-0-string-1"
      inline={false}
      isInvalid={false}
      isValid={false}
      key="0"
      label="Option One"
      onChange={[Function]}
      title=""
      type="checkbox"
    >
      <div
        className="custom-control custom-checkbox"
      >
        <FormCheckInput
          as="input"
          checked={true}
          disabled={false}
          isInvalid={false}
          isStatic={false}
          isValid={false}
          onChange={[Function]}
          type="checkbox"
        >
          <input
            checked={true}
            className="custom-control-input"
            disabled={false}
            id="check-0-string-1"
            onChange={[Function]}
            type="checkbox"
          />
        </FormCheckInput>
        <FormCheckLabel
          title=""
        >
          <label
            className="custom-control-label"
            htmlFor="check-0-string-1"
            title=""
          >
            Option One
          </label>
        </FormCheckLabel>
      </div>
    </FormCheck>
    <FormCheck
      checked={false}
      custom={true}
      disabled={false}
      id="check-1-number-2"
      inline={false}
      isInvalid={false}
      isValid={false}
      key="1"
      label="Option Two"
      onChange={[Function]}
      title=""
      type="checkbox"
    >
      <div
        className="custom-control custom-checkbox"
      >
        <FormCheckInput
          as="input"
          checked={false}
          disabled={false}
          isInvalid={false}
          isStatic={false}
          isValid={false}
          onChange={[Function]}
          type="checkbox"
        >
          <input
            checked={false}
            className="custom-control-input"
            disabled={false}
            id="check-1-number-2"
            onChange={[Function]}
            type="checkbox"
          />
        </FormCheckInput>
        <FormCheckLabel
          title=""
        >
          <label
            className="custom-control-label"
            htmlFor="check-1-number-2"
            title=""
          >
            Option Two
          </label>
        </FormCheckLabel>
      </div>
    </FormCheck>
    <FormCheck
      checked={false}
      custom={true}
      disabled={false}
      id="check-2-string-thiisareallylongstringvalue"
      inline={false}
      isInvalid={false}
      isValid={false}
      key="2"
      label="Option Three"
      onChange={[Function]}
      title=""
      type="checkbox"
    >
      <div
        className="custom-control custom-checkbox"
      >
        <FormCheckInput
          as="input"
          checked={false}
          disabled={false}
          isInvalid={false}
          isStatic={false}
          isValid={false}
          onChange={[Function]}
          type="checkbox"
        >
          <input
            checked={false}
            className="custom-control-input"
            disabled={false}
            id="check-2-string-thiisareallylongstringvalue"
            onChange={[Function]}
            type="checkbox"
          />
        </FormCheckInput>
        <FormCheckLabel
          title=""
        >
          <label
            className="custom-control-label"
            htmlFor="check-2-string-thiisareallylongstringvalue"
            title=""
          >
            Option Three
          </label>
        </FormCheckLabel>
      </div>
    </FormCheck>
  </div>
</Component>
`;

exports[`It matches the snapshot with title and description 1`] = `
<Component
  onSelectChange={[Function]}
  options={
    Array [
      Object {
        "label": "Option One",
        "value": "1",
      },
      Object {
        "label": "Option Two",
        "value": 2,
      },
      Object {
        "label": "Option Three",
        "value": "thiisareallylongstringvalue",
      },
    ]
  }
  value={
    Array [
      "1",
      2,
    ]
  }
>
  <div>
    <h4
      className="mb-1"
    />
    <p
      className="text-muted mb-2 mt-1"
      hidden={true}
    />
    <FormCheck
      checked={true}
      custom={true}
      disabled={false}
      id="check-0-string-1"
      inline={false}
      isInvalid={false}
      isValid={false}
      key="0"
      label="Option One"
      onChange={[Function]}
      title=""
      type="checkbox"
    >
      <div
        className="custom-control custom-checkbox"
      >
        <FormCheckInput
          as="input"
          checked={true}
          disabled={false}
          isInvalid={false}
          isStatic={false}
          isValid={false}
          onChange={[Function]}
          type="checkbox"
        >
          <input
            checked={true}
            className="custom-control-input"
            disabled={false}
            id="check-0-string-1"
            onChange={[Function]}
            type="checkbox"
          />
        </FormCheckInput>
        <FormCheckLabel
          title=""
        >
          <label
            className="custom-control-label"
            htmlFor="check-0-string-1"
            title=""
          >
            Option One
          </label>
        </FormCheckLabel>
      </div>
    </FormCheck>
    <FormCheck
      checked={false}
      custom={true}
      disabled={false}
      id="check-1-number-2"
      inline={false}
      isInvalid={false}
      isValid={false}
      key="1"
      label="Option Two"
      onChange={[Function]}
      title=""
      type="checkbox"
    >
      <div
        className="custom-control custom-checkbox"
      >
        <FormCheckInput
          as="input"
          checked={false}
          disabled={false}
          isInvalid={false}
          isStatic={false}
          isValid={false}
          onChange={[Function]}
          type="checkbox"
        >
          <input
            checked={false}
            className="custom-control-input"
            disabled={false}
            id="check-1-number-2"
            onChange={[Function]}
            type="checkbox"
          />
        </FormCheckInput>
        <FormCheckLabel
          title=""
        >
          <label
            className="custom-control-label"
            htmlFor="check-1-number-2"
            title=""
          >
            Option Two
          </label>
        </FormCheckLabel>
      </div>
    </FormCheck>
    <FormCheck
      checked={false}
      custom={true}
      disabled={false}
      id="check-2-string-thiisareallylongstringvalue"
      inline={false}
      isInvalid={false}
      isValid={false}
      key="2"
      label="Option Three"
      onChange={[Function]}
      title=""
      type="checkbox"
    >
      <div
        className="custom-control custom-checkbox"
      >
        <FormCheckInput
          as="input"
          checked={false}
          disabled={false}
          isInvalid={false}
          isStatic={false}
          isValid={false}
          onChange={[Function]}
          type="checkbox"
        >
          <input
            checked={false}
            className="custom-control-input"
            disabled={false}
            id="check-2-string-thiisareallylongstringvalue"
            onChange={[Function]}
            type="checkbox"
          />
        </FormCheckInput>
        <FormCheckLabel
          title=""
        >
          <label
            className="custom-control-label"
            htmlFor="check-2-string-thiisareallylongstringvalue"
            title=""
          >
            Option Three
          </label>
        </FormCheckLabel>
      </div>
    </FormCheck>
  </div>
</Component>
`;
